Project.toml

```diff
@@ -1,11 +1,12 @@
 name = "NNlib"
 uuid = "872c559c-99b0-510c-b3b7-b6c96a88d5cd"
-version = "0.8.20"
+version = "0.9.0"
 
 [deps]
 Adapt = "79e6a3ab-5dfb-504d-930d-738a2a938a0e"
 Atomix = "a9b6321e-bd34-4604-b9c9-b65b8de01458"
 ChainRulesCore = "d360d2e6-b24c-11e9-a2a3-2a2ae2dbcce4"
+cuDNN = "02a925ec-e4fe-4b08-9a7e-0d78e3d38ccd"
```

Comment:
I had to make `cuDNN` a strong dependence because it seems that there is no way to make `using CUDA` also trigger the loading of cuDNN from the extension. This is not ideal but the alternatives seem to be worse:
- have the extension triggered by `using CUDA, cuDNN`
- keep `NNlibCUDA` as a separate package

Reply:
Is there some other option?

Reply:
Making it a strong dep is a non-starter because `cuDNN` directly depends on CUDA.jl. I would say we go with option 1, but there are a couple variations on it we could also consider:
1. Make only `cuDNN` a weak dep and access CUDA.jl through `cuDNN.CUDA`.
2. Create separate extensions for CUDA.jl and `cuDNN`. Then someone can choose to only load the former if they don't need e.g. conv functionality. This doesn't make anything easier so we should probably consider it after an extension is in place.

Reply:
I really don't like `using Flux, CUDA, cuDNN`, it seems a huge shame to have to load two packages, one of which is a super-obscure internal thing. I mean I don't even know where it lives,  https://www.google.com/search?q=cuDNN.jl leads me only to one abandoned 5 years ago. 

It's a huge shame to give up on `using Flux, CUDA` as the interface. I understand that the default use of new package extensions does not allow us to then load cuDNN. I wonder if we should seriously consider either hacks for now (can Requires or something like it load cuDNN?) or finding out if upstream can be changed e.g. to unify cuDNN into CUDA.

Reply:
> 2. Create separate extensions for CUDA.jl and cuDNN. Then someone can choose to only load the former if they don't need e.g. conv functionality. This doesn't make anything easier so we should probably consider it after an extension is in place.

One nice thing about this option is that if you go `using Flux, CUDA` and then run a convolution, we can print a friendly error telling you about this obscure package `] add cuDNN`. Then "doing the obvious thing" of `using Flux, CUDA` will guide you to the non-obvious thing, if necessary.

Reply:
> It's a huge shame to give up on using Flux, CUDA as the interface. I understand that the default use of new package extensions does not allow us to then load cuDNN. I wonder if we should seriously consider either hacks for now (can Requires or something like it load cuDNN?) or finding out if upstream can be changed e.g. to unify cuDNN into CUDA.

I agree here. Doing `using Flux, CUDA, cuDNN` is sufficiently convoluted to be confusing. While separate extensions does mean better errors, it still means we have something complex for the most common use case.

Can we not create in FluxML (or have JuliaGPU do this) an umbrella package that has CUDA and cuDNN as hard dependencies? Then this is a weak dep of NNlib? Users would now have `using Flux, FluxCUDA`. Perhaps this is more confusing. We would also still have the release cycle annoyances, but now the package is trivial so bumping the version should be easy.

Reply:
To be clear I think 90% of my objection to using three things applies to `using Flux, cuDNN`. It's an obscure package whose most obvious version has been dead for years. I presume it was intended for "internal" use, but the way that extensions ended up happening pushes us towards "exposing" it. The user-facing package is CUDA.jl.

Reply:
cuDNN is actually its own package now again, https://github.com/JuliaGPU/CUDA.jl/tree/master/lib/cudnn

Reply:
I think Michael's point is that people are conditioned to search for some variant of "<package name>.jl" or "julia <package name>", neither of which turns up the current subpackage right now. Not sure if it'd be possible to de-index the results that do show up or publish a docs page for https://github.com/JuliaGPU/CUDA.jl/tree/master/lib/cudnn (and ensure it rises to the top).

Reply:
Making the path from the old package to the new one super-clear would be great. Like a giant yellow box on the README, and archive the repository, or something?

But IMO is still seems a bit awful if the way you load Flux+CUDA is instead to load this other weird package nobody has heard about. Which is a collection of internal methods basically for hooking NNlib (which is a sort-of internal part of Flux) up to arrays etc. from CUDA.jl. I don't have a great solution but I'm keen not to bake in a really weird public interface, just because of how the pieces were sitting when package extensions landed.

I wonder a bit whether cuDNN ought to be something like CUDANNlibExt? Maybe that's too narrow as it may have other users.  Could the same source code file be included both in the package cuDNN.jl and some CUDANNlibExt like this? Maybe not...

Reply:
> Alternatively, perhaps the various CUDA.jl subpackages could be reintegrated as [Rust-style package features](https://doc.rust-lang.org/cargo/reference/features-examples.html)? I'm not sure if we'd get something similar using Preferences.jl

With Preferences.jl, it would be rather clunky. AFAICT, it would require users to depend on CUDA.jl in order to activate cuDNN, which doesn't seem much better than `using cuDNN`.

Features as a first-class thing would seem nice, but probably not achievable in the near future?

Reply:
> it would require users to depend on CUDA.jl in order to activate cuDNN, which doesn't seem much better than `using cuDNN`.

Don't users have to depend on CUDA.jl either way since it's a hard dep of cuDNN? I confess I can't imagine a case where one would depend on the latter without also depending on the former.

Reply:
Some packages might want to expose GPU-accelerated functionality without users having to depend on either CUDA or CUDNN. With Preferences, the user environment would then need to include CUDA (i.e. in the Project.toml) in order to set the CUDNN preference.

Reply:
Does https://github.com/JuliaPackaging/Preferences.jl/issues/24 work for package extension dependencies? The example seems to imply that packages can set preferences for their dependencies.

Reply:
I'm not sure. In any case, that example also shows how the active project needs to have `A` as a hard dependency, either in `[deps]` or in `[extras]`, which is the point I was making above.

That said, although I'm happy to consider alternative ways of loading CUDNN-like functionality, I don't see it happening soon. Without a first-class language feature and using Preferences.jl, it would require users to import CUDA.jl to enable the CUDNN features, which IMO is mostly the same as having them do `using cuDNN`. And even with a first-class feature where, say, packages could express in their Project.toml which features they request of a package, it doesn't seem clear how that would interact with package extensions (what if a user has CUDA.jl but not CUDA.jl+CUDNN, etc).

Reply:
By sheer coincidence, I noticed https://github.com/JuliaPackaging/Preferences.jl/issues/53 was reported today.

Reply:
FWIW, I'm currently too busy working on other things to consider reworking the CUDA.jl/cuDNN.jl situation again (especially now that it just stabilized a bit after introducing JLLs), but I'm not opposed to changes. So if anybody would want to explore a different mechanism for shipping CUDA libraries, feel free to open an issue or PR.